src/runtime/local/io/WriteMM.h

```diff
@@ -0,0 +1,239 @@
+            field = MM_INT_STR;
+        } else if (std::is_floating_point<VT>::value) {
+            field = MM_REAL_STR;
+        } else {
+            throw std::runtime_error("WriteMM::apply: Unsupported data type");
+        }
+
+        const char *symmetry = MM_GENERAL_STR;
+
+        f << MatrixMarketBanner << " matrix " << format << " " << field << " " << symmetry << std::endl;
+
+        size_t rows = arg->getNumRows();
+        size_t cols = arg->getNumCols();
+        size_t nnz = countNNZ(arg, symmetry);
+
+        f << rows << " " << cols << " " << nnz << std::endl;
+
+        const size_t *rowOffsets = arg->getRowOffsets();
+        const size_t *colIdxs = arg->getColIdxs();
+        const VT *values = arg->getValues();
+
+        std::vector<std::vector<std::pair<size_t, VT>>> colEntries(cols);
+
+        for (size_t i = 0; i < rows; ++i) {
+            for (size_t idx = rowOffsets[i]; idx < rowOffsets[i + 1]; ++idx) {
+                size_t j = colIdxs[idx];
+                VT val = values[idx];
+                colEntries[j].emplace_back(i, val);
+            }
+        }
+
+        if (strcmp(symmetry, MM_GENERAL_STR) == 0) {
+            for (size_t j = 0; j < cols; ++j) {
+                for (const auto &entry : colEntries[j]) {
+                    size_t i = entry.first;
+                    VT val = entry.second;
+                    if (strcmp(field, MM_REAL_STR) == 0) {
+                        if (val >= 0) {
+                            f << i + 1 << " " << j + 1 << "  " << std::scientific << std::setprecision(13) << val
+                              << std::endl;
+                        } else {
+                            f << i + 1 << " " << j + 1 << " " << std::scientific << std::setprecision(13) << val
+                              << std::endl;
+                        }
+                    } else {
+                        f << i + 1 << " " << j + 1 << " " << val << std::endl;
+                    }
+                }
+            }
+        } else {
+            throw std::runtime_error("WriteMM::apply: Unsupported symmetry type");
+        }
+
+        f.close();
+    }
+
+  private:
+    static size_t countNNZ(const CSRMatrix<VT> *arg, const char *symmetry) {
+        size_t nnz = 0;
+        size_t rows = arg->getNumRows();
+        size_t cols = arg->getNumCols();
+
+        std::vector<std::vector<std::pair<size_t, VT>>> colEntries(cols);
+
+        const size_t *rowOffsets = arg->getRowOffsets();
+        const size_t *colIdxs = arg->getColIdxs();
+        const VT *values = arg->getValues();
+
+        for (size_t i = 0; i < rows; ++i) {
+            for (size_t idx = rowOffsets[i]; idx < rowOffsets[i + 1]; ++idx) {
+                size_t j = colIdxs[idx];
+                colEntries[j].emplace_back(i, values[idx]);
+            }
+        }
+
+        if (strcmp(symmetry, MM_GENERAL_STR) == 0) {
+            nnz = arg->getNumNonZeros();
+        } else {

```

Comment:
I think I don't fully understand why the computation above is needed if  you return getNumNonZeros or fail.